std-lib/list/list.lurk

```diff
@@ -0,0 +1,66 @@
+;; a list is of the form (cons a (cons b (cons ... (cons nil)...)))
+
+;; list : make a list out of a (possibly unevaluated) list quote
+(letrec ((list (lambda (q)
+		 (if (car q)
+		     (cons (eval (car q)) (list (cdr q)))
+		     '()))))
+  (current-env))
+
+;; list-eq : determine if two lists are equal
+(letrec ((list-eq (lambda (l1 l2)
+    (if (car l1) 
+        (if (car l2)
+            (if (= (car l1) (car l2))
+                (list-eq (cdr l1) (cdr l2))
+                nil)
+            nil)
+        (if (car l2)
+            nil
+            t)))))
+    (current-env))
```

Comment:
Shouldn't this whole file be one big `letrec`? As it stands, aren't we returning a bunch of different envs, each with one function in it? Maybe I'm misunderstanding what `(current-env)` does.

Reply:
Also what is the difference between `list-eq` and `eq`?

```
> (eq '(a (b)) '(a (b)))
[3 iterations] => T
```

Reply:
haha, that's funny, somehow I thought that eq didn't work on `bool` and on lists, but it works on both! These are removed

Reply:
re it being one big `letrec` or several, that's a good question. I don't know what convention is; as I understand it, as written each `let` and `letrec` updates the current environment to include that variable binding. I've only bundled related/helper functions into the same `let` or `letrec`. @porcuquine thoughts on this stylistically?

Reply:
Maybe (current-env) doesn't do what I thought.

I guess it is not only returning the env at that point, it's also setting the top level's state to include all the variables from that point. In that case the name is a little misleading, the name sounds like it doesn't mutate any state. I'd call it something like `capture-env`, or `use-env` or `set-top-level-env` etc.

Reply:
`(current-env)` returns the current lexical environment.

Separately, when a `.lurk` file is processed by `!(:load …)`, the toplevel lexical environment used by the REPL is replaced with the result of evaluating each form encountered.

This has the net effect that sequential forms, each extending the environment, can be used to build up an environment containing all the definitions. This trick allows us to define libraries in different files than use them, and for libraries to depend on other libraries. It also allows more modular definitions within a single file.

This is mostly a hack that won't be so important once the RAM subset is available, but it certainly makes things easier in the current world, which we do want and need to support.

Reply:
So @porcuquine what's your thought on whether to define the lib in one big letrec or a bunch of different ones? I lean toward just using one (assuming that would work just as well).

Reply:
Probably multiple, but logically grouped. I'm not 100% sure, but that's my instinct. That gives us an extra layer of structure and will probably help as intermediate stage toward breaking sections out into their own file/module/whatever. It will also facilitate re-ordering for performance.